### PR TITLE
Add Soromox semi-implicit Euler integration

### PR DESCRIPTION
**Pull request by mstoelzle**
## Summary

- add a Diffrax-compatible, system-bound `SemiImplicitEuler` solver for
  Soromox `[q, qd, auxiliary]` states
- extend the Section IVb parallel-rollout benchmark with solver selection and
  explicit finite-state diagnostics
- validate one shared solver/timestep across ArticulatedSoftRobot, PlanarPCS,
  PCS, and GVS models with 1, 2, 4, 8, 16, and 32 segments
- retain `Tsit5` with `sim_dt=1e-4` as the benchmark default because it was
  stable throughout the matched sweep and 40.3% cheaper in aggregate than
  `SemiImplicitEuler` with `sim_dt=1e-5`
- document standard Diffrax solver support, adaptive step-size controllers,
  and floating-base quaternion integration assumptions

### Soromox semi-implicit Euler

`soromox.simulation.SemiImplicitEuler` implements a deterministic first-order
kick--drift update through the Diffrax `AbstractSolver` interface. It evaluates
the complete Soromox derivative once, updates `qd`, and retracts the resulting
velocity increment onto configuration space before repacking the state.

The solver is bound to a `DynamicalSystem` and uses its `split_state`,
`pack_state`, and `retract_configuration` helpers. It therefore supports
unequal spatial floating-base configuration and velocity dimensions,
quaternion retraction, and trailing system auxiliary state rather than relying
on an equal half-split. Diffrax-level control and environment leaves retain an
explicit-Euler update.

The regression suite covers the kick--drift ordering, JIT and `vmap`, a
spatial floating-base PCS state with unequal dimensions, quaternion
normalization, and a floating PlanarHSA hysteresis state.

### Benchmark solver and stability reporting

The Section IVb generator now accepts `--solver` with `euler`,
`semi-implicit-euler`, `heun`, `bosh3`, `tsit5`, and `dopri5`. The selected
solver is shared across every requested system and size; `tsit5` remains the
default.

Each CSV row records the solver, whether the final rollout state is finite, and
the maximum absolute final-state value. Non-finite rollouts are also marked in
the terminal output so unstable timings cannot be mistaken for useful
throughput.

### Diffrax and floating-base documentation

The new simulation-integrator API page links to Diffrax's solver and step-size
controller catalogs and shows both adaptive `Tsit5` and fixed-step
`SemiImplicitEuler` usage. It documents that standard Diffrax solvers support
the complete Soromox first-order state and receive projected quaternions before
dynamics evaluations and on saved output, while still advancing quaternion
storage in ambient coordinates rather than retracting every Runge--Kutta stage.

## GPU rollout evaluation

The matched comparison used an NVIDIA RTX 5090 in FP64 with a 1 s horizon,
batch size 4, perturbation scale `1e-3`, seed 0, no warmup, and one timed repeat
for all four model families and all six segment counts. These are cold
end-to-end timings and therefore include JIT compilation equally for both
configurations.

| Shared configuration | Finite cases | Aggregate wall time | Relative cost |
| --- | ---: | ---: | ---: |
| `Tsit5`, `sim_dt=1e-4` | 24/24 | 1517.86 s | 1.000x |
| `SemiImplicitEuler`, `sim_dt=1e-5` | 24/24 | 2542.47 s | 1.675x |

All 96 perturbed environments were finite for both configurations. The
semi-implicit configuration was 1.58--1.74x slower in every paired case, with
a median slowdown of 1.667x. Separate limiting-case probes found non-finite
GVS behavior for `Tsit5` at `2e-4` and non-finite long-horizon PCS behavior for
`SemiImplicitEuler` at `2.5e-5`; the selected steps are respectively 2x and
2.5x smaller than those tested failing points.

## Validation

- focused integrator and benchmark suite: `26 passed`
- Ruff lint and formatting checks on every changed Python file: passed
- Zensical documentation build: passed
- `git diff --check`: passed
- complete matched GPU sweeps: 24/24 finite cases for each configuration

### CODE REVIEW

File: paper_results/secIVb_parallel_rollouts_gpu/code/generate_parallel_rollout_benchmark.py [hunk 1]
```diff
@@ -572,6 +620,13 @@ def _run_benchmarks(args: argparse.Namespace, device: jax.Device) -> int:
                         repeats=args.repeats,
                     )
                     ts_last, *_ = outputs
+                    state_outputs = outputs[1:]
+                    rollout_finite = bool(
+                        all(jnp.all(jnp.isfinite(value)) for value in state_outputs)
+                    )
+                    max_abs_state = float(
+                        max(jnp.max(jnp.abs(value)) for value in state_outputs)
+                    )
```
[the remark this answers is not in the document]
Comment (mstoelzle) [reply paper_results/secIVb_parallel_rollouts_gpu/code/generate_parallel_rollout_benchmark.py]: Addressed in 2033280. Maximum absolute state magnitude is now computed with a stacked JAX reduction and converted only after the shared jax.device_get call; Python max is no longer used on JAX scalars.

File: paper_results/secIVb_parallel_rollouts_gpu/code/generate_parallel_rollout_benchmark.py [hunk 2]
```diff
@@ -572,6 +620,13 @@ def _run_benchmarks(args: argparse.Namespace, device: jax.Device) -> int:
                         repeats=args.repeats,
                     )
                     ts_last, *_ = outputs
+                    state_outputs = outputs[1:]
+                    rollout_finite = bool(
+                        all(jnp.all(jnp.isfinite(value)) for value in state_outputs)
+                    )
```
[the remark this answers is not in the document]
Comment (mstoelzle) [reply paper_results/secIVb_parallel_rollouts_gpu/code/generate_parallel_rollout_benchmark.py]: Addressed in 2033280. The new _rollout_state_diagnostics helper stacks and reduces all finiteness checks with JAX, then transfers the combined finiteness/max result to the host through one jax.device_get call. The regression also asserts the single host-transfer boundary.

File: src/soromox/simulation/integrators.py [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+        Returns:
+            A Diffrax step tuple containing the state at ``t1``, no local error
+            estimate, linear-interpolation data, no persistent solver state,
+            and a successful result code.
+        """
+        del solver_state, made_jump
+        control = terms.contr(t0, t1)
+        increment = terms.vf_prod(t0, y0, args, control)
+        y1 = jax.tree.map(lambda initial, delta: initial + delta, y0, increment)
```
[the remark this answers is not in the document]
Comment (mstoelzle) [reply src/soromox/simulation/integrators.py]: Addressed in 2033280. JAX's default PyTree handling already treats None as a structural node, so the existing open-loop rollout did not actually evaluate None plus None; nevertheless, the solver now makes the contract explicit by treating None as a leaf and preserving it. A regression verifies absent control and environment states remain None.
